Validate sym traceability

marcosps opened the pull request:
Hey guys, this is something that I studies recently, and it would help to detect some corner cases of livepatch creation. The possibillity of having the livepatch procedure to be even more automatic makes such improvement even more important, since we can catch these issues right away.

Please take a look and say what you think about it!

marcosps committed:
klplib: codestream: Use get_symbol_by_name

By using get_symbol_by_name the code can be simplified, instead of always
getting all the symbols and creating a cache for all the symbols per
arch/codestream.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps commented on klpbuild/klplib/codestream.py:
```diff
@@ -368,12 +369,46 @@ def lp_out_file(self, lp_name, fname):
+        for i in range(0, len(sec_data), 8):
+            ptr = sec_data[i:i+8]
+
+            data = int.from_bytes(ptr, byteorder="little")
```
True! This was required when I tried to print the addresses. I'll check if this is necessary for us to adjust depending on the arch, so keep this on hold for me. I'm planning to extend the test to all archs so we can validate this properly.

vmezzela commented:
Hi Marcos, thanks for the PR! I have one question, as far as I can see with this PR also symbols to be externalized are being checked. Is that intended? Should we check the traceability of the affected symbols only?

marcosps commented:
> Hi Marcos, thanks for the PR! I have one question, as far as I can see with this PR also symbols to be externalized are being checked. Is that intended? Should we check the traceability of the affected symbols only?

That's true, it was not the intention. I was so focused on the setup that I forgot that this function is also called on extraction. I'll change it the code to only check this on setup phase. Good catch!

marcosps committed:
extract: Stop checking for -pg compile flag

The next commit will prohibit to livepatch a symbol that is not patchable,
so this test will fail since this check is done at extract, while the
symbol checking will be done at setup phase.

Remove the test now since it'll be obsolete in the next patches.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
tests/supported.csv: Update the list of codestreams

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
klplib: ibs: Always extract all rpm packages for all architectures

If a given livepatch only affects a non x86 architecture the klp-build
tool will fail to find the ipa-clones files and other files that are
present on kernel-default-devel. Since the biggest rpm was kernel-source,
and we are not downloading and extracting it anymore, just extract
everything else like we do for x86.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
klplib: codestream: Use arch to find obj and the ipa clones files

klp-build always try to find the files using the x86 architecture as part
of the path, but this doesn't work for livepatches that affect only
architectures different from x86. To fix it always search for the obj
and ipa clones using the first of the affected architectures on setup.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
file2config: Check if a given file is only compiled on one arch

There is a corner case of when a file being patched lives inside arch,
it would wrongly says that all other architectures are affected, for
example, arch/s390/kernel/perf_cpum_sf.c. The current code will find
the CONFIG_PERF_EVENTS entry, and tell that all archs are affected, when
in fact only s390x is.

This is not the complete solution, given that a specific CONFIG_$ARCH
must be checked with the additional config found, but we may extend this
functionality when such corner case is found.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
templ: Use cs.get_default_archs when checking if IS_ENABLED should be used

The current check doesn't work on codestreams that are enabled on only
a subset of the supported architectures, like MICRO (which is enabled
only on x86 and s390x).

This happens because the function checks for all ARCHS against the affected
ones. To fix it compare the affected architectures with the architectures
supported on the codestream.

Also, sort the entries of the default_archs method since it's used
on Templ class to compare with the affected one, and also sort the
archs passed by arguments on setup.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
codestream: Add check for symbol being patchable

Kernel code can add instructions to the compiler do not allow the symbol
to be traceable/profiled, which also disabled the livepatching of the symbol.

Such cases can be tricky to found, and usually discovered only in the test
phase of a livepatch development. This change introduces a way to detect
such symbols by looking at the ftrace addresses created on build time.

For now it only checks the symbols on vmlinux. To extend the same solution
to kernel modules would involve resolving relocation manually, which is tricky.

Also, ppc64le is different regarding the content of start_mcount_loc and
stop_mcount_loc, since it gets an address pointing to function descriptors,
so the code code checks for the range of the symbol + symbol size.

With this new check some tests had to change their testing symbols, since
some of them are not really patchable.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
tests: Bump codestream versions of some tests

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps commented:
> > @vmezzela I believe that now we are covered! And also I removed an argument from another method that seems unnecessary now. Would you mind taking a second look?
> 
> Sure, I added a comment. Sorry, I just spotted it now.
> 
> Thank for dropping that argument, I think it was indeed unnecessary :)

Well, it seems that it was not. We need to check for the archs that are affected by the CVE, and the current loop over cs.archs is related to the archs supported on the codestream... Maybe we can add a comment to clarify that :)

marcosps commented:
The new commits in this branch are necessary for those cases where the livepatch doesn't affect x86, and so setup or extract could fail, since we always check things against x86.

marcosps committed:
utills: Add preferred_arch

This new function will return one architecture affected by the given CVE.
The preference is always x86_64, then s390x and then ppc64le. This is
necessary so we can call klp-ccp with an architecture different from
x86_64 when necessary.

Adapt all Codestream methods that used ARCH by default to use preferred_arch
instead, making klp-build capable of managing CVEs that do not affect x86.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
plugins: extract: Remove s390x specific compiler arguments

klp-cco can't handle them, and they only affect code generation.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
extract: Drop klp-ccp comments when comparing code

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
klplib: data: Only call download_cs_data if there are missing codestreams

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
plugins: setup: Do not print message if no codestreams are affected

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
plugins: extract: Drop __init/__exit attributes

klp-ccp currently keeps these attributes, and often generates them
along with the symbol, which is wrong. To resolve this issue just remove
the attributes.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
tests: templ: Fix typo when trying to test only one arch

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps commented:
@fgyanz this is now ready, would you mind reviewing it again? It's passing all tests.

fgyanz commented on klpbuild/plugins/extract.py:
```diff
@@ -806,7 +806,7 @@ def start_extract(lp_name, lp_filter, no_patches, avoid_ext):
         for obj, syms in obj_syms.items():
-            missing = cs.check_symbol_archs(get_codestreams_data('archs'), obj, syms, True)
+            missing = cs.check_symbol_archs(get_codestreams_data('archs'), obj, syms, True, False)
             if missing:
```
I think it's safe to remove this guard. The function returns {}, which can .items().

marcosps replied:
done

fgyanz commented on klpbuild/klplib/codestream.py:
```diff
@@ -382,24 +384,100 @@ def lp_out_file(self, lp_name, fname):
 
-            elif len(syms) > 1:
+            if len(syms) > 1:
                 print(f"WARNING: {self.full_cs_name()}-{arch} ({self.kernel}): symbol {symbol} duplicated on {mod}")
```
Do you mind mind changing this print to logging.warning() ?

marcosps replied:
done

fgyanz commented on klpbuild/klplib/codestream.py:
```diff
@@ -382,24 +384,100 @@ def lp_out_file(self, lp_name, fname):
+
+        pos = bisect.bisect_left(trace_addrs, val)
+        if pos == len(trace_addrs) or trace_addrs[pos] != val:
+            raise RuntimeError(f"{self.full_cs_name()}-{arch} ({self.kernel}): Symbol {name} has tracing disabled.")
```
Would it be possible to use logging.error and exit() instead of RuntimeError? The latter is too verbose.
Same for the rest of RuntimeErrors

marcosps replied:
done

fgyanz commented on klpbuild/plugins/extract.py:
```diff
@@ -270,6 +270,7 @@ def get_klpp_symbols(out_dir, lp_out):
                 logging.warning(f"Failed to find klpp_{sym} in {lp_out}")
                 continue
             klpp_proto = re.sub(r'\s+',' ', m.group(2)).strip() + ';'
+            klpp_proto = re.sub(r' __(init|exit)', ' ', klpp_proto)
```
Do you mind adding a simple comment clarifying why this is needed? Same below :)

marcosps replied:
Done

fgyanz reviewed (commented):
Hi @marcosps , the PR looks great! Just some minor nitpicking here and there.

One last thing, would it be possible to raise a big `WARNING` at the end of `extract` command in case a user decides to generate livepatches for a non-x86_64 arch? 
The reason being that `klp-ccp`  does not currently support other archs, so generated livepatches won't probably work. Even worse, they might compile and run, but underneath the code might be wrong. I think we have the responsibility of alerting the user in such case :)

Thanks a lot for the good work :partying_face:

fgyanz commented on klpbuild/klplib/file2config.py:
```diff
@@ -97,6 +97,17 @@ def find_configs_for_files(cs, file_paths: list):
+                configs[path] = {'conf': "CONFIG_X86_64", 'module': 'vmlinux'}
+            elif "powerpc" in path:
+                configs[path] = {'conf': "CONFIG_PPC64", 'module': 'vmlinux'}
+            continue
```
I think we can make it a bit more robust by doing it just after the `_find_config()` and:
```suggestion
        if path.startswith("arch") and len(cs.get_all_configs(conf)) > 1:
            if "s390" in path:
                configs[path] = {'conf': "CONFIG_S390", 'module': 'vmlinux'}
            elif "x86" in path:
                configs[path] = {'conf': "CONFIG_X86_64", 'module': 'vmlinux'}
            elif "powerpc" in path:
                configs[path] = {'conf': "CONFIG_PPC64", 'module': 'vmlinux'}
            continue
```
So if the found `conf` is too generic, fallback to the arch specific one. What do you think? :)

marcosps replied:
done

marcosps commented:
@fgyanz ready for another revision :)

marcosps committed:
plugins: extract: Add warning for non x86 livepatches

As klp-ccp doesn't support other architectures besides x86, add a warning
to the developer when this is the case.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

fgyanz reviewed (approved):
Thanks a lot for the changes! LGTM!
